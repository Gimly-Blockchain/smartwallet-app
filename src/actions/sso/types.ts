import { PublicProfileClaimMetadata } from 'cred-types-jolocom-core/types'

/**
 * @dev Simply using all claims required by the public profile
 */

export type IssuerPublicProfileSummary = PublicProfileClaimMetadata['claimInterface']

/**
 * @dev An identity summary is composed of a DID + all public info (currently public profile)
 */
export interface IdentitySummary {
  did: string
  publicProfile?: IssuerPublicProfileSummary
}

export interface AuthenticationRequestSummary {
  callbackURL: string
  requester: IdentitySummary
  description: string
  requestJWT: string
}