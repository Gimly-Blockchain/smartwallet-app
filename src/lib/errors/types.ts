import { routeList } from 'src/routeList'
import { ErrorCode } from './codes'

export { ErrorCode }

export interface IAppError {
  origError: any
  navigateTo: routeList
}

export interface UserReport {
  userError: string | undefined
  userDescription: string
  userContact: string
  sendPrivateData: boolean
}

export interface ErrorReport extends UserReport {
  error: IAppError | Error | undefined
}
