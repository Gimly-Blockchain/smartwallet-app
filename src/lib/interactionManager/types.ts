
import { JSONWebToken, JWTEncodable } from 'jolocom-lib/js/interactionTokens/JSONWebToken'


/* Responsabilities:
 * 1. continue an interaction
 */
export interface IInteractionManager {
    tokenRecieved: (token: JSONWebToken<JWTEncodable>) => void

}
