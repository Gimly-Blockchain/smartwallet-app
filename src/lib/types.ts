import { JSONWebToken, JWTEncodable } from 'jolocom-lib/js/interactionTokens/JSONWebToken'

export type SendFn = (token: JSONWebToken<JWTEncodable>) => Promise<any>
