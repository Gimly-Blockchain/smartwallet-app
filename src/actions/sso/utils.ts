import { IssuerPublicProfileSummary } from './types'
import { Identity } from 'jolocom-lib/js/identity/identity'

export const parsePublicProfile = (identity: Identity) => {
  const { publicProfile, did } = identity
  if (!publicProfile) {
    return {
      did,
    }
  }
  const { id, ...parsedProfile } = publicProfile.claim
  return {
    did,
    publicProfile: parsedProfile as IssuerPublicProfileSummary,
  }
}
