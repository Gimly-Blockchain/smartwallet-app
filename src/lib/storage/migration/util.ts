export async function getPassword(): Promise<string> {
  // At the time of writing this file, the encryption password is stored in
  // react-native-keychain and accessesed through src/lib/keychain.ts, but code
  // is reproduced here in case things change in the future

  if (typeof navigator !== 'undefined' && navigator.product == 'ReactNative') {
    // We load react-native-keychain here conditionally because it is not
    // transpiled and cannot be loaded into ts-node (if running migrations locally
    // on dev machine)
    const Keychain = require('react-native-keychain')

    const keyChainData = await Keychain.getGenericPassword()
    if (keyChainData && keyChainData.password) {
      return keyChainData.password
    } else {
      throw new Error("Can't load password from react-native-keychain")
    }
  } else {
    const password = process.env.SMARTWALLET_PASSWORD
    if (!password) {
      throw new Error(
        'Node envrionment detected. ' +
          'Please set a password in the environment variable SMARTWALLET_PASSWORD',
      )
    }
    return password
  }
}
