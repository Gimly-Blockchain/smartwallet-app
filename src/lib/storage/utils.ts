// TODO rework
// Multiple credentials of the same type, i.e. ['Credential', 'ProofOfNameCredential']
import {
  CredentialEntity,
  VerifiableCredentialEntity,
} from './entities'

interface ModifiedCredentialEntity {
  propertyName: string
  propertyValue: string[]
  verifiableCredential: VerifiableCredentialEntity
}

export const groupAttributesByCredentialId = (
  credentials: CredentialEntity[],
): ModifiedCredentialEntity[] => {
  // Convert values to arrays for easier concatination later
  const modifiedAttributes = credentials.map(credential => ({
    ...credential,
    propertyValue: [credential.propertyValue], // "eugeniu@jolocom.com" -> ["eugeniu@jolocom.com"]
  }))

  // Helper function
  const findByCredId = (
    arrToSearch: ModifiedCredentialEntity[],
    value: ModifiedCredentialEntity,
  ) =>
    arrToSearch.findIndex(
      entry => entry.verifiableCredential.id === value.verifiableCredential.id,
    )

  debugger
  return modifiedAttributes.reduce(
    (acc: ModifiedCredentialEntity[], curr: ModifiedCredentialEntity) => {
      const matchingIndex = findByCredId(acc, curr)

      // This seems to be a hack, uniting individual claims from the credential back together
      if (matchingIndex >= 0) {
        acc[matchingIndex].propertyValue = [
          ...acc[matchingIndex].propertyValue,
          ...curr.propertyValue,
        ]
        return acc
      } else {
        return [...acc, curr]
      }
    },
    [],
  )
}
