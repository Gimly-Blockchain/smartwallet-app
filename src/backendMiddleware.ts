import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { EncryptionLib, EncryptionLibInterface } from 'src/lib/crypto'
import { Storage } from 'src/lib/storage/storage'
import { KeyChain, KeyChainInterface } from 'src/lib/keychain'
import { ConnectionOptions } from 'typeorm/browser'
import {
  createJolocomRegistry,
  JolocomRegistry,
} from 'jolocom-lib/js/registries/jolocomRegistry'
import { IpfsCustomConnector } from './lib/ipfs'
import { jolocomContractsAdapter } from 'jolocom-lib/js/contracts/contractsAdapter'
import { jolocomEthereumResolver } from 'jolocom-lib/js/ethereum/ethereum'
import { jolocomContractsGateway } from 'jolocom-lib/js/contracts/contractsGateway'
import { JolocomLib } from 'jolocom-lib'
import { KeyTypes } from 'jolocom-lib/js/vaultedKeyProvider/types'
import { publicKeyToDID } from 'jolocom-lib/js/utils/crypto'
import { Identity } from 'jolocom-lib/js/identity/identity'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { generateSecureRandomBytes } from './lib/util';

export enum ErrorCodes {
  NoEntropy = 'NoEntropy',
  NoKeyProvider = 'NoKeyProvider',
  NoWallet = 'NoWallet',
  DecryptionFailed = 'DecryptionFailed',
}

export class BackendError extends Error {
  static codes = ErrorCodes

  constructor(code: ErrorCodes) {
    super(code)
  }
}

export class BackendMiddleware {
  private _identityWallet!: IdentityWallet
  private _keyProvider!: SoftwareKeyProvider
  private _entropyData!: { encryptedEntropy: string; timestamp: number }

  public storageLib: Storage
  public encryptionLib: EncryptionLibInterface
  public keyChainLib: KeyChainInterface
  public registry: JolocomRegistry

  public constructor(config: {
    fuelingEndpoint: string
    typeOrmConfig: ConnectionOptions
  }) {
    this.storageLib = new Storage(config.typeOrmConfig)
    this.encryptionLib = new EncryptionLib()
    this.keyChainLib = new KeyChain()
    this.registry = createJolocomRegistry({
      ipfsConnector: new IpfsCustomConnector({
        host: 'ipfs.jolocom.com',
        port: 443,
        protocol: 'https',
      }),
      ethereumConnector: jolocomEthereumResolver,
      contracts: {
        adapter: jolocomContractsAdapter,
        gateway: jolocomContractsGateway,
      },
    })
  }

  public async initStorage(): Promise<void> {
    await this.storageLib.initConnection()
  }

  public get identityWallet(): IdentityWallet {
    if (this._identityWallet) return this._identityWallet
    throw new BackendError(ErrorCodes.NoWallet)
  }

  public get keyProvider(): SoftwareKeyProvider {
    if (this._keyProvider) return this._keyProvider
    throw new BackendError(ErrorCodes.NoKeyProvider)
  }

  public get entropyData(): BackendMiddleware['_entropyData'] {
    if (this._entropyData) return this._entropyData
    throw new BackendError(ErrorCodes.NoEntropy)
  }

  public async prepareIdentityWallet(): Promise<IdentityWallet> {
    if (this._identityWallet) return this._identityWallet

    const encryptedEntropy = await this.storageLib.get.encryptedSeed()
    if (!encryptedEntropy) throw new BackendError(ErrorCodes.NoEntropy)

    this._entropyData = { encryptedEntropy, timestamp: Date.now() }
    const encryptionPass = await this.keyChainLib.getPassword()

    const decryptedSeed = this.encryptionLib.decryptWithPass({
      cipher: this._entropyData.encryptedEntropy,
      pass: encryptionPass,
    })

    if (!decryptedSeed) {
      throw new BackendError(ErrorCodes.DecryptionFailed)
    }

    // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
    this._keyProvider = JolocomLib.KeyProvider.fromSeed(
      Buffer.from(decryptedSeed, 'hex'),
      encryptionPass,
    )

    const userPubKey = this._keyProvider.getPublicKey({
      derivationPath: KeyTypes.jolocomIdentityKey,
      encryptionPass,
    })

    const didDocument = await this.storageLib.get.didDoc(
      publicKeyToDID(userPubKey),
    )

    if (didDocument) {
      const identity = Identity.fromDidDocument({ didDocument })

      // TODO Simplify constructor
      return (this._identityWallet = new IdentityWallet({
        identity,
        vaultedKeyProvider: this._keyProvider,
        publicKeyMetadata: {
          derivationPath: KeyTypes.jolocomIdentityKey,
          keyId: identity.publicKeySection[0].id,
        },
        contractsAdapter: this.registry.contractsAdapter,
        contractsGateway: this.registry.contractsGateway,
      }))
    } else {
      const { jolocomIdentityKey: derivationPath } = JolocomLib.KeyTypes
      const identityWallet = await this.registry.authenticate(
        this._keyProvider,
        {
          encryptionPass,
          derivationPath,
        },
      )

      await this.storageLib.store.didDoc(identityWallet.didDocument)
      return (this._identityWallet = identityWallet)
    }
  }

  public async setEntropy(encodedEntropy: string): Promise<void> {
    const password = (await generateSecureRandomBytes(32)).toString('base64')
    const encEntropy = this.encryptionLib.encryptWithPass({
      data: encodedEntropy,
      pass: password,
    })
    this._entropyData = { encryptedEntropy: encEntropy, timestamp: Date.now() }

    // TODO: rework the seed param on lib, currently cleartext seed is being passed around. Bad.
    this._keyProvider = JolocomLib.KeyProvider.fromSeed(
      Buffer.from(encodedEntropy, 'hex'),
      password,
    )
  }

  public async fuelKeyWithEther(): Promise<void> {
    const password = await this.keyChainLib.getPassword()
    await JolocomLib.util.fuelKeyWithEther(
      this.keyProvider.getPublicKey({
        encryptionPass: password,
        derivationPath: JolocomLib.KeyTypes.ethereumKey,
      }),
    )
  }

  public async createIdentity(): Promise<Identity> {
    const password = await this.keyChainLib.getPassword()
    this._identityWallet = await this.registry.create(
      this.keyProvider,
      password,
    )

    const personaData = {
      did: this._identityWallet.identity.did,
      controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
    }
    await this.storageLib.store.persona(personaData)
    await this.storageLib.store.didDoc(this._identityWallet.didDocument)
    await this.storageLib.store.encryptedSeed(this._entropyData)
    await this.keyChainLib.savePassword(password)

    return this._identityWallet.identity
  }
}
