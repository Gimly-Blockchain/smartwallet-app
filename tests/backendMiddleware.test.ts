import * as util from 'src/lib/util'
import { BackendMiddleware, BackendError } from 'src/backendMiddleware'
import { stub, reveal } from './utils'
import { ConnectionOptions } from 'typeorm/browser'
import data from 'tests/actions/registration/data/mockRegistrationData'
import { JolocomLib } from 'jolocom-lib'
import { IRegistry } from 'jolocom-lib/js/registries/types'

const MockDate = require('mockdate')

const mockBackendMiddlewareConfig = {
  fuelingEndpoint: 'hptp://petrol1.station',
  typeOrmConfig: {} as ConnectionOptions,
}

describe('BackendMiddleware', () => {
  const { getPasswordResult, cipher, entropy, identityWallet } = data
  const keyChainLib = stub<BackendMiddleware['keyChainLib']>({
    getPassword: jest.fn().mockResolvedValue(getPasswordResult)
  })
  const decryptWithPass = jest.fn().mockReturnValue(entropy)
  const storageLib = {
    store: stub<BackendMiddleware['storageLib']['store']>(),
    get: stub<BackendMiddleware['storageLib']['get']>({
      encryptedSeed: jest.fn().mockResolvedValue(cipher),
    }),
  }
  const registry = stub<IRegistry>()

  const createBackendMiddleware = (): BackendMiddleware => {
    const backendMiddleware = new BackendMiddleware(mockBackendMiddlewareConfig)
    Object.assign(backendMiddleware, {
      keyChainLib,
      encryptionLib: { decryptWithPass },
      storageLib,
      registry,
    })
    return backendMiddleware
  }

  jest
    .spyOn(util, 'generateSecureRandomBytes')
    .mockImplementation(() =>
      Promise.resolve(Buffer.from(getPasswordResult, 'base64')),
    )

  describe('prepareIdentityWallet', () => {
    const backendMiddleware = createBackendMiddleware()

    it('should throw NoEntropy if there is no stored entropy', async () => {
      reveal(storageLib.get).encryptedSeed.mockResolvedValueOnce(null)
      const walletPromise = backendMiddleware.prepareIdentityWallet()
      return expect(walletPromise).rejects.toThrow(BackendError.codes.NoEntropy)
    })

    it('should throw DecryptionFailed if decryption fails', async () => {
      decryptWithPass.mockReturnValueOnce(null)
      const walletPromise = backendMiddleware.prepareIdentityWallet()
      return expect(walletPromise).rejects.toThrow(
        BackendError.codes.DecryptionFailed,
      )
    })

    it('should authenticate and cache the identity if not cached', async () => {
      reveal(storageLib.get).didDoc.mockResolvedValueOnce(null)
      reveal(registry).authenticate.mockResolvedValue(identityWallet)

      await backendMiddleware.prepareIdentityWallet()
      expect(reveal(registry).authenticate.mock.calls).toMatchSnapshot()
      expect(reveal(storageLib.get).didDoc.mock.calls).toMatchSnapshot()
      expect(storageLib.store.didDoc).toHaveBeenCalledWith(
        identityWallet.didDocument,
      )
      expect(() => backendMiddleware.identityWallet).not.toThrow()
    })

    it('should use cached identity if available', async () => {
      stub.clearMocks(registry)
      stub.clearMocks(storageLib.store)
      reveal(storageLib.get).didDoc.mockResolvedValueOnce(
        identityWallet.didDocument,
      )

      await backendMiddleware.prepareIdentityWallet()
      expect(reveal(registry).authenticate).not.toHaveBeenCalled()
      expect(storageLib.store.didDoc).not.toHaveBeenCalled()

      expect(() => backendMiddleware.identityWallet).not.toThrow()
      expect(backendMiddleware.identityWallet.identity.did).toMatch(
        identityWallet.didDocument.did,
      )
    })
  })

  describe('Identity Creation', () => {
    const backendMiddleware = createBackendMiddleware()
    let entropyData: { encryptedEntropy: string; timestamp: number }

    beforeAll(() => {
      MockDate.set(new Date(946681200000))
      entropyData = {
        encryptedEntropy: cipher,
        timestamp: Date.now(),
      }
    })

    afterAll(() => MockDate.reset())

    it('should throw NoKeyProvider from fuelKeyWithEther if no entropy was set', async () => {
      const fuelPromise = backendMiddleware.fuelKeyWithEther()
      return expect(fuelPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should throw NoKeyProvider from createIdentity if no entropy was set', async () => {
      const identityPromise = backendMiddleware.createIdentity()
      return expect(identityPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should throw NoKeyProvider from fuelKeyWithEther if no entropy was set', async () => {
      const fuelPromise = backendMiddleware.fuelKeyWithEther()
      return expect(fuelPromise).rejects.toThrow(
        BackendError.codes.NoKeyProvider,
      )
    })

    it('should createKeyProvider', async () => {
      await backendMiddleware.createKeyProvider(entropy)
      expect(() => backendMiddleware.keyProvider).not.toThrow()
      expect(backendMiddleware.keyProvider['encryptedSeed'].toString('base64')).toMatch(cipher)
    })

    it('should not store anything before the identity is registered', () => {
      expect(storageLib.store.didDoc).not.toHaveBeenCalled()
      expect(storageLib.store.persona).not.toHaveBeenCalled()
      expect(storageLib.store.encryptedSeed).not.toHaveBeenCalled()
    })

    it('should fuelKeyWithEther', async () => {
      const fuelSpy = jest
        .spyOn(JolocomLib.util, 'fuelKeyWithEther')
        .mockResolvedValueOnce(null)

      await backendMiddleware.fuelKeyWithEther()
      expect(fuelSpy.mock.calls).toMatchSnapshot()
    })

    it('should register the identity and store data on createIdentity', async () => {
      reveal(registry).create.mockResolvedValue(identityWallet)

      const identity = await backendMiddleware.createIdentity()

      expect(registry.create).toHaveBeenCalledWith(
        backendMiddleware.keyProvider,
        getPasswordResult,
      )
      expect(storageLib.store.didDoc).toHaveBeenCalledWith(
        identityWallet.didDocument,
      )
      expect(storageLib.store.persona).toHaveBeenCalledWith({
        did: identity.did,
        controllingKeyPath: JolocomLib.KeyTypes.jolocomIdentityKey,
      })
      expect(keyChainLib.savePassword).toHaveBeenCalledTimes(1)
      expect(keyChainLib.savePassword).toHaveBeenCalledWith(
        getPasswordResult,
      )

      expect(storageLib.store.encryptedSeed).toHaveBeenCalledWith(entropyData)
    })
  })
})
