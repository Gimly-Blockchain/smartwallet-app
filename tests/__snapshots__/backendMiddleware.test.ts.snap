// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackendMiddleware Identity Creation should fuelKeyWithEther 1`] = `
Array [
  Array [
    Object {
      "data": Array [
        2,
        190,
        222,
        236,
        223,
        239,
        223,
        32,
        17,
        81,
        201,
        162,
        78,
        254,
        96,
        10,
        97,
        136,
        138,
        150,
        254,
        179,
        2,
        244,
        3,
        231,
        162,
        235,
        71,
        152,
        56,
        126,
        92,
      ],
      "type": "Buffer",
    },
  ],
]
`;

exports[`BackendMiddleware prepareIdentityWallet should authenticate and cache the identity if not cached 1`] = `
Array [
  Array [
    SoftwareKeyProvider {
      "encryptedSeed": Object {
        "data": Array [
          183,
          116,
          81,
          20,
          81,
          84,
          212,
          61,
          81,
          173,
          61,
          49,
          234,
          209,
          98,
          116,
          146,
          8,
          172,
          20,
          172,
          74,
          237,
          124,
          18,
          0,
          151,
          9,
          149,
          229,
          236,
          165,
        ],
        "type": "Buffer",
      },
    },
    Object {
      "derivationPath": "m/73'/0'/0'/0",
      "encryptionPass": "hunter0=",
    },
  ],
]
`;

exports[`BackendMiddleware prepareIdentityWallet should authenticate and cache the identity if not cached 2`] = `
Array [
  Array [
    "did:jolo:d129fdce5ff7d69878c001b0973d7e1a0ff2ae4cc7066d5a769e909c28c4149d",
  ],
]
`;
